### ISSUE TITLE
Spawning in hundreds/thousands of visual.Box in SceneCanvas

### ISSUE THREAD

**Issue by bobzwik** [vispy/vispy#1814]
Hello,

This is more of a help request topic rather than an issue. I'm trying to render thousands of visual.box objects in a SceneCanvas, and I'm running into some major performance issues (because I clearly don't know what I'm doing).

Some context:
To learn Python, I've built a [quadcopter simulation](https://github.com/bobzwik/Quadcopter_SimCon/tree/potentialField) and I'm slowly adding things to it. In the current state of the project, the quadcopter navigates through a pointcloud tunnel using a potential field ([see youtube video](https://youtu.be/goYXPfm020w)). I'm using Markers in a SceneCanvas to display the pointcloud, and as you can see, I'm updating the colors of the markers depending or their distance to the drone.

However, I would like to change these markers for boxes. That also change color. 

I've played around with the following code, which spawns a line of 100 boxes, and changes the color of half of them every 0.05 s (20 fps).
```
import sys
import numpy as np
from vispy import scene
from vispy.color import Color
from vispy import app
from vispy.visuals.transforms import STTransform

# Pointcloud
pc_size = 100

point_x = np.arange(pc_size)
point_y = np.zeros(pc_size)
point_z = np.zeros(pc_size)

pointcloud = np.array([point_x, point_y, point_z]).T

# Canvas
canvas = scene.SceneCanvas(keys='interactive', size=(800, 600), show=True)
canvas.measure_fps()

# Set up a viewbox to display the cube with interactive arcball
view = canvas.central_widget.add_view()
view.bgcolor = '#efefef'
view.camera = scene.cameras.TurntableCamera()
view.padding = 100

color = Color((1,0,0))

cubes = [scene.visuals.Box(1, 1, 1, color=color, edge_color="black", parent=view.scene) for i in range(len(pointcloud))]
for cube, i in zip(cubes, range(len(cubes))):
    transform = STTransform(translate=pointcloud[i])
    cube.transform = transform

red = True
idx_to_change = np.arange(1,len(cubes),2)

def update(ev):
    global cubes, red, idx_to_change
    if (red):
        for cube in [cubes[i] for i in idx_to_change]:
            cube.mesh.color = Color((0,1,0))
        red = False
    else:
        for cube in [cubes[i] for i in idx_to_change]:
            cube.mesh.color = Color((1,0,0))
        red = True


timer = app.Timer(0.05)
timer.connect(update)
timer.start(iterations=100)

if __name__ == '__main__' and sys.flags.interactive == 0:
    canvas.app.run()
```

The output framerate is however 2.7 fps. And the more I add boxes, the slower it gets (to the point of crashing when I load my 24000 point pointcloud).

So basically, I asking if what I'm trying to do is even possible with VisPy. And if not, can anything in Python handle it? I know my computer can, but with ROS and RViz.

If you think VisPy can somehow handle this, I'm really interested in learning how.

Thanks!

Also: 
```
Platform: Windows-10-10.0.18362-SP0
Python:   3.7.6 (tags/v3.7.6:43364a7ae0, Dec 19 2019, 00:42:30) [MSC v.1916 64 bit (AMD64)]
NumPy:    1.18.1
Backend:  PyQt5
pyqt4:    None
pyqt5:    ('PyQt5', '5.14.1', '5.14.0')
pyside:   None
pyside2:  None
pyglet:   None
glfw:     None
sdl2:     None
wx:       None
egl:      None
osmesa:   None
_test:    None
```

**Comment by djhoese** [comment]
Thanks for asking this question. It comes up pretty often (I should really add something to the documentation that I can point to).

Every Visual object in VisPy represents a single OpenGL program. GL programs are run serially by your GPU and when redrawing a scene (typically) you need to redraw *every* program so that every element you want to draw is drawn to the screen. What can happen (and is happening to you) is that you have a lot of Visuals which means you have a lot of GL programs. This means your GPU has to draw *every* one of those objects, one at a time, before it can draw the next "frame". As you can imagine this results in terrible performance. Even worse, you are using a Box visual which is actually a CompoundVisual made up of two MeshVisuals; so 2x as many GL programs.

The solution is to use Visuals, whether builtin to VisPy or made custom by you, that draw multiple objects in one program. Good examples of this are the LineVisual and the MarkersVisual. In your case I would look at the MarkersVisual as a starting point and see if you can get at least 90% of the functionality you're looking for. If this doesn't meet all your needs then you could look in to sub-classing it and customizing it the way you need.

Another future option are the "collection" objects that were ported from glumpy (I think), but currently aren't compatible with the Visual interface (they use the low-level `gloo` interface). I'm working on getting funding that might give me some time to make these compatible with Visuals and the SceneCanvas.

If you go the subclassing route let me know and we can talk about the specifics.

**Comment by bobzwik** [comment]
Thanks for the fast reply!

Yeah, [I currently have it working with scene.visuals.Marker()](https://youtu.be/goYXPfm020w) (is there a difference between using that and visuals.MarkerVisual?). I get 16 fps (strange fps behavior using app.TImer, might do start another issue for that), which 100% fits my debugging needs, but not quite my aesthetics need (for the long term).

I am interested in the subclassing method, but don't know enough Python to see how this will help. When making my animation, I opted to go the Scene and Visual route instead of the gloo route, because it was important for me to have a turntable camera (I think it's possible with gloo, just more complicated, right?).

Another solution I can imagine is to create a whole mesh of the pointcloud, so there would be just 1 GL object (or more depending on the number of colors I'm using). I'm a bit apprehensive of going this way.

Thanks again!

**Comment by djhoese** [comment]
The `vispy.visuals.XVisual`objects are meant for the base Canvas object. They don't understand the parent/child and transforms logic that the SceneCanvas uses. For that, use the `vispy.scene.visuals.X`. These objects are just a simple wrapper around the `vispy.visuals.XVisual` objects.

For subclassing, think if the Markers class did almost everything you needed and you looked at the source code and though "if only I could change this one part of this class to do what I want". With subclassing you could make your own class, override the method or two that you need to modify, but reuse all the base Markers functionality. You are right on the visual versus gloo stuff. Visuals are built on the gloo objects and as I described above, the SceneCanvas visuals understand the necessary concepts to deal with cameras in a nice way.

Depending on exactly what you're doing with the point cloud and how you are updating it, I would suggest having one Marker instance and update the data shown with it based on the current information you have. If you are drawing *a lot* of data points then you could look in to grouping the markers in to a few objects and only drawing the ones that are needed (the GPU does this automatically...usually). You could also look at some of the other visuals like the Mesh or Volume which may perform better at drawing large shapes but might be limited for the exact experience you want.

**Comment by bobzwik** [comment]
Alright, thanks for the clear explanation. I'll look into subclassing eventually, but I'll probably stay with Markers for now.

Currently, I have 1 Marker instance (```scatter```). Every frame, I create an array of colors of the same size as the pointcloud, then I reset the data and input the color array with ```scatter.set_data(pointcloud, face_color=colors)```. Is that a decent enough method?

**Comment by bobzwik** [comment]
There are some performance (fps) issues with this method, but if I created a subclass of Markers with a ```set_color``` method, where I'd just set the ```data['a_bg_color']```  in [here](https://github.com/vispy/vispy/blob/master/vispy/visuals/markers.py) to a new color array, do you think I would gain performance?

I'd also have to re-run :
```
self._vbo.set_data(data)
self.shared_program.bind(self._vbo)

self.update()
```

right?

**Comment by djhoese** [comment]
That should work. I don't remember though, do you have to provide the pointcloud data every time or can you set the colors by themselves? If not available now, there may be a PR waiting to be merged that has this functionality.

As for performance issues with this method, it really depends what is slowing you down. Are you creating the colors fast? How big is that array? If you subclassed and you held on to that array structure and only updated the colors then that *should* work. You should only have to run `self.update()` after modifications.

if you went even further and customized the shader, if you can use something about the position to determine the color and let the GPU calculate it that would be by far the fastest.

**Comment by bobzwik** [comment]
Yeah, currently ```set_data``` has a default value of ```None```, so I imagine that I have to provide the pointcloud data everytime (I know that I have to set ```size``` and ```edge_color``` everytime, otherwise it takes the default values).

Here is set_data : 

```
def set_data(self, pos=None, symbol='o', size=10., edge_width=1.,
                 edge_width_rel=None, edge_color='black', face_color='white',
                 scaling=False):
```

I'm running my simulation first, accumulating data, and then I start the animation. During the simulation, I create 3 arrays of booleans of size (1,n), where n is the number of points in the pointcloud (around 24000 points). I accumulate these 3 arrays over time to finally create 3 m by n arrays, where m is the number of timesteps simulated.

When comes the time to animate each frame, I use the boolean arrays to create the color array, like so:
```
colors[np.where(notInRange_all[i,:])[0]] = np.array([1, 1, 0, 0.7])
colors[np.where(inRange_all[i,:])[0]] = np.array([0, 1, 0, 0.7])
colors[np.where(inField_all[i,:])[0]] = np.array([1, 0, 0, 0.7])
scatter.set_data(pointcloud, edge_color=None, face_color=colors, size=6)
```
where ```i``` is incremented each frame by a number depending of the desired framerate. I think this is one of the most efficient ways to create the color array, but I'm open to suggestions!

**Comment by djhoese** [comment]
I really thought there was a pull request for this, but now I can't find it.

So based on that `colors` code, are you saying there is a for loop around that block of code? Or is `i` updated with each drawn frame?

**Comment by bobzwik** [comment]
I have ```i``` as a nonlocal variable. At first, I setup the canvas, then I set a timer connected to a ```update()``` method.

Basically:

```
## Code to setup Canvas ##

i = 0

def update(ev):
       nonlocal i
       ## Calculate how to draw 2 lines representing the quadcopter, and draw it ##

        view.camera.azimuth = view.camera.azimuth-psi_diff
        view.camera.center = [x,y,z]

        colors[np.where(notInRange_all[i*numFrames,:])[0]] = np.array([1, 1, 0, 0.7])
        colors[np.where(inRange_all[i*numFrames,:])[0]] = np.array([0, 1, 0, 0.7])
        colors[np.where(inField_all[i*numFrames,:])[0]] = np.array([1, 0, 0, 0.7])
        scatter.set_data(pointcloud, edge_color=None, face_color=colors, size=6)
    
        i += 1

timer = app.Timer()
timer.connect(update)
timer.start(iterations=numFrames)
```
As you can see, on top of setting the Marker data every time the update() function is called, I move the camera to follow slightly behind the quadcopter. I've confirmed that moving the camera does greatly affect performance. I think my main issue is with that, but I'll try seeing what a ```set_color```
method can do.

**Comment by djhoese** [comment]
I'm not sure how much it will effect performance, but you could also look at subclassing the SceneCanvas so you can store those nonlocal variables and other objects (colors, timer, etc) as instance attributes instead of globals.

The other thing that could be happening is the canvas may be doing multiple redraws for each update. One for the azimuth change, one ofr the center change, and one for the set_data. There are ways to avoid this, but I've had mixed results when trying to do it in my own work (EventEmitters have ways of blocking update signals).

**Comment by bobzwik** [comment]
I've conducted a quick test:

- Animation with camera movement, colors update and scatter.set_data : 
    - Average 25-27 FPS
- Animation with camera movement, but without colors update and but with scatter.set_data : 
    - Average 27-28 FPS
- Animation with camera movement and colors update but without scatter.set_data : 
    - Average 29-31 FPS
- Animation with camera movement, but without colors update and without scatter.set_data : 
    - Average 31-33 FPS
- Animation without camera movement, but with colors update and scatter.set_data : 
    - Average 45-48 FPS

To get these results, I set the timer to run at 1 kHz (I don't know another way to force it to run as fast as possible).

So the camera movements affect performance more, probably because of the reason you mentioned.  I then It's the ```set_data```, and then the colors update.

I'm interested in the ways to block each draw (center change, azimuth change and ```set_data```). In the future, I won't be using app.Timer, I'm building myself my own scheduler for parallel processing (essentially, I'd have a process running the quad dynamics, one the controller and one the animation). I'll have the animation process running in a while loop, looping as fast as it can (if that's possible).

**Comment by djhoese** [comment]
I'm a little busy this week so I'm not sure I can help immediately with the event blocking stuff. You might have to search the source code to make a guess at how to do it.

As for the multi-process method, the GUI application typically has to be run in the main thread. If you build a Qt GUI you could use a separate QThread that gets the new animation data, then uses Qt signals to notify the main GUI thread that it needs to update and provide the new data. That way the GUI thread isn't polling/looping.

**Comment by bobzwik** [comment]
> As for the multi-process method, the GUI application typically has to be run in the main thread. If you build a Qt GUI you could use a separate QThread that gets the new animation data, then uses Qt signals to notify the main GUI thread that it needs to update and provide the new data. That way the GUI thread isn't polling/looping.

Oh my, that goes way over my head. I've never touched anything Qt. I might try, but maybe I'll keep the animation in the main process and have the scheduler in a different process.

Just to be sure, I'm not talking about threads, I'm talking about different processes, running on different cores.

**Comment by djhoese** [comment]
> Just to be sure, I'm not talking about threads, I'm talking about different processes, running on different cores. Is that what you were thinking too?

That's fine, just that threads are generally easier to pass data around with and my go to way of communicating in a parallel fashion with GUIs. Most of my experience is with Qt and it is one of the more heavily tested backends for VisPy, that's the only reason I mentioned it.

**Comment by bobzwik** [comment]
Hey there David,
I've taken some time to play around with my code, here's my progress so far.

I've create a subclass of visuals.Markers, with a `set_color` method, like so:
`class ColorMarkers(visuals.Markers):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
    
    def set_color(self, face_color='white'):
        face_color = ColorArray(face_color).rgba
        if len(face_color) == 1:
            face_color = face_color[0]

        self._data['a_bg_color'] = face_color
        self._vbo.set_data(self._data)
        self.update()
`
With this new class, I went from 25-27 fps to 27-29 fps. Not much, but it's a start.

Like I said before, I was trying out these fps measurements by setting the app.Timer very fast (0.001s). But I wanted to eventually use a different method of animating my simulation.

Instead of fixing a framerate, I want the animation to run as fast as possible, and display the right frame corresponding to elapsed time. I tried this with time.perf_counter(). Basically :

```
startTime = time.perf_counter()

def update():
        nonlocal idx_prev
        currentTime = time.perf_counter()-startTime

        # t_all is my array of all simulated times, argmax returns the first nonzero element, but returns 0 when currentTime is greater than all values of t_aa
        idx_now = np.argmax(t_all > currentTime)                  
        if (idx_now == 0 and idx_prev > 0):
            return (False, idx_prev)
        else:
             
            ## Update frame using index idx_now
            view.camera.azimuth = view.camera.azimuth-psi_diff
            view.camera.center = [x,y,z]
            colors[np.where(notInRange_all[idx_now,:])[0]] = np.array([1, 1, 0, 0.7])
            colors[np.where(inRange_all[idx_now,:])[0]] = np.array([0, 1, 0, 0.7])
            colors[np.where(inField_all[idx_now,:])[0]] = np.array([1, 0, 0, 0.7])
            scatter.set_color(face_color=colors)

            canvas.app.process_events()
            idx_prev = idx_now

        return (True, idx_prev)

runAni = True
idx_prev = 0

while runAni:
        runAni, idx_prev = update(idx_prev)
```

This works like a charm, and I'm now getting 50 fps!!! And if I comment out the lines to set the center and azimuth to follow the drone, I get 100 fps!!

Now, I wanted to look into event blocking, but my knowledge on events in really limited (I've just begun playing with Python six months ago on my spare time), so I quickly gave up. However, in the camera classes, the property setter for azimuth and center finish by calling the `self.view_changed` method. I decided to write a subclass of the turntable camera, override those 2 property setters and leave out that last line, like so : 

```
class NonUpdatingTurntable(vispy.scene.cameras.TurntableCamera):
    def __init__(self, **kwargs):
        super(NonUpdatingTurntable, self).__init__(**kwargs)
    
    @property
    def azimuth(self):
        """ The angle of the camera in degrees around the y axis. An angle of
        0 places the camera within the (y, z) plane.
        """
        return self._azimuth

    @azimuth.setter
    def azimuth(self, azim):
        azim = float(azim)
        while azim < -180:
            azim += 360
        while azim > 180:
            azim -= 360
        self._azimuth = azim
        # self.view_changed()
    
    @property
    def center(self):
        """ The center location for this camera

        The exact meaning of this value differs per type of camera, but
        generally means the point of interest or the rotation point.
        """
        return self._center or (0, 0, 0)
    
    @center.setter
    def center(self, val):
        if len(val) == 2:
            self._center = float(val[0]), float(val[1]), 0.0
        elif len(val) == 3:
            self._center = float(val[0]), float(val[1]), float(val[2])
        else:
            raise ValueError('Center must be a 2 or 3 element tuple')
        # self.view_changed()
```

And then I call ```view.camera.view_changed``` at the end of my update() function. I therefore update the camera only one. And the result is .... 70 fps!! Way more than enough.

Thank you for your help! Eventually I might look into changing the markers for cubes. We'll see.

**Comment by bobzwik** [comment]
There's however a catch.

These results were done on my laptop. When I run the same code on my desktop however, I get a solid cap at 20.0 fps. No matter what I try. I've tried the same things as every test I've presented in my few posts. A solid 20.0 fps. Would you have an idea why?

My laptop is a 4ish year old laptop:
- i5 6300HQ
- 8 GB RAM
- Nvidia GTX 950M (and I've made sure that Python applications run on my dedicated GPU, and not the Intel one)

My desktop is less than a year old, with much better specs : 
- AMD Ryzen 5 2600
- 16 GB RAM
- AMD Radeon RX 580 (It's hard to tell if my GPU is being used though. Both AMD's System Monitor and Windows Task Manager report that Python uses 3-6% GPU when the animation is rolling (0% when it's not).

I can't believe that my hardware can't handle vispy animations if my laptop can. Is there a parameter I have to play with somewhere for AMD GPUs, either in Vispy or Radeon settings?

I have read online that AMD doesn't run openGL as well as Nvidia, but I feel like that 20 fps cap is software. What do you think?

Thanks again!

**Comment by bobzwik** [comment]
Sorry, literally 15 seconds after my last comment, I looked at Radeon Settings. I found this : 
![Untitled](https://user-images.githubusercontent.com/5113091/74001136-73eb0600-4938-11ea-8572-f11dc89da4ac.png)

It was set to "Balanced". I set it to "Optimize Performance" and guess what... 95-100 fps for my animation. Better than my laptop, as expected.

**Comment by djhoese** [comment]
Wow, what an exciting read to wake up to! Glad you got things working! If you have any suggestions or recommendations for VisPy to make a workflow like this easier, that would be great. Pull requests for any changes you think would help are very welcome.

**Comment by bobzwik** [comment]
Yeah, I was thinking of doing a PR for the `set_color` method, but stopping the camera from updating at each set of a parameter is going to be trickier.

I however still think that my 100.0 fps on my AMD machine is software limited. Something in Vispy maybe has to be changed to tell my system to use my GPU at 100% ? I mean, my video games weren't capped at 20 fps when I was in balanced mode. Anyways, I'm happy with my little patch!

**Comment by djhoese** [comment]
The "right" solution for the camera stuff might be for us to properly document (maybe update an example too) with how to block signals for things like this, if possible.

As for performance versus video games, you may be hitting the limits of what Python and VisPy can do for you. I'm sure there are ways the code could be updated to make it perform even better but at some point there is a limit and I'm not sure Python on top of OpenGL is going to get you that 240Hz gaming monitor-type behavior.

I would still recommend looking at subclassing the SceneCanvas and storing things internal to the class. Also, do you need the `canvas.app.process_events()` call? That is usually considered bad practice unless you know you need it. Usually a `canvas.update()` is used instead.

**Comment by bobzwik** [comment]
> I would still recommend looking at subclassing the SceneCanvas and storing things internal to the class.

Yeah, I should try that too. Storing the camera, lines and scatter in a subclass of SceneCanvas, and writing methods to update them depending on my drone's position and color array.

> Also, do you need the canvas.app.process_events() call? That is usually considered bad practice unless you know you need it. Usually a canvas.update() is used instead.

`canvas.update()` didn't work for me, the canvas remains white and hangs for the duration of the animation. `process_events()` did the trick for me. Could it be because I removed the `app.run()` that's normally at the end of the script (I removed it because my animation now just runs with the `while` loop).

**Comment by djhoese** [comment]
>  Could it be because I removed the app.run() that's normally at the end of the script (I removed it because my animation now just runs with the while loop).

Oh yes, absolutely! The application should be running, but now I'm scared that you'll find out there are needless updates/draws that aren't being blocked and should be. Or maybe it will run even faster. Hard to say.

**Comment by bobzwik** [comment]
I tried re-adding the `app.run()` command, but because my `update()` function doesn't act on events, my animation doesn't run. 

I went back to the `app.Timer()` method (instead of the while loop). I'm getting similar fps when I don't set the framerate (but when I do, performance really dips (?)). I kinda like my "while loop" technique, and nothing wrong seems to be happening. I might try looking into generating events with my while loop, but I'm quite lost with events. Do you know of a quick way to implement events?

Thanks!

**Comment by djhoese** [comment]
The preferred way of doing this is to use a Timer with `'auto'` (`0` might also work) as the interval along with using `app.run()`. If you were to use other GUI components besides just the VisPy canvas in your application you would run in to serious issues managing when draws and user interactions happen.

**Comment by bobzwik** [comment]
Alright, so to avoid problems, I went back using `app.run()` . I'm now looking at creating a subclass of visual.Markers to create cubes instead of 2D symbols. And I'm stumped. I've found where the symbol function is decided (line 611 of ```markers.py```) Here is a snippet of that code.

```
    @symbol.setter
    def symbol(self, symbol):
        if symbol == self._symbol:
            return
        if (symbol is not None and self._symbol is None and
                self._data is not None):
            # Allow user to configure symbol after a set_data call with
            # symbol=None. This can break down if the user does a consecutive
            # marker.symbol = 'disc'
            # marker.symbol = None
            # without drawing. At this point the memory leaking ensues
            # but this case is unlikely/makes no sense.
            self._vbo.set_data(self._data)
            self.shared_program.bind(self._vbo)
        self._symbol = symbol
        if symbol is None:
            self._marker_fun = None
        else:
            _check_valid('symbol', symbol, marker_types)
            self._marker_fun = Function(_marker_dict[symbol])
            self._marker_fun['v_size'] = self._v_size_var
            self.shared_program.frag['marker'] = self._marker_fun
        self.update()
```

I would have to set ```self._marker_fun``` to a different function. Do I find a way to draw cubes the same way the other symbols are drawn (they're only drawn in 2D though), do I use ```visuals.Box()```? Or do I create a new function that projects cubes depending in 2D depending my camera's position, view angle and fov?

Thanks

**Comment by djhoese** [comment]
Oh boy. You'd probably want to define a new GLSL shader function similar to the existing marker functions. But to do it in 3D...I'm not sure how well that would work. Theoretically it shouldn't be a problem, but I'm not sure what it would take.

You wouldn't use visuals.Box unless you are just taking parts of the Box shader code to implement the function above.

**Comment by bobzwik** [comment]
So I found the GLSL shader function in [MeshVisual (line 99)](https://github.com/vispy/vispy/blob/master/vispy/visuals/mesh.py), and if I understand correctly, I just need the functions that doesn't have any special shaders:

```# Shader code for non lighted rendering
vertex_template = """
varying vec4 v_base_color;

void main() {
    v_base_color = $color_transform($base_color);
    gl_Position = $transform($to_vec4($position));
}
"""

fragment_template = """
varying vec4 v_base_color;
void main() {
    gl_FragColor = v_base_color;
}
"""
```

where `$transform` is the transform mapping of the visual object coordinate system to the 2D render coordinate system ([define at line 523](https://github.com/vispy/vispy/blob/master/vispy/visuals/mesh.py)) and `$position` is the vertices.

Am I on the right track? I basically have to find a way to implement this in a custom subclass of MarkersVisual, right?

Thanks!

**Comment by djhoese** [comment]
I was thinking you would make a new function like:

https://github.com/vispy/vispy/blob/5e2985a8a0ec34e2353a1885cc4eb39ab97c0aec/vispy/visuals/markers.py#L184-L191

And use that instead of the existing ones in the `markers.py` module. As you may know the MarkersVisual lets you change the symbol of what is drawn on the screen. I was hoping you could add a new function for a cube marker...maybe.

You are mostly correct on the transform/position stuff. A key point is that `$transform` is actually a dynamically generated shader function in Vispy's SceneCanvas that is a chaining of all of the transforms from Screen -> SceneCanvas -> View -> Child Node -> ... -> Visual object. Each one of these elements has or can have their own transform. VisPy's TransformSystem chains these together as a single `$transform` function. You likely don't want to modify that too much since that determines where on the screen an element shows up.

**Comment by djhoese** [comment]
Hhhmm I'm realizing now though that this is only sending you a 2D coordinate. That probably has to be updated for your to properly draw the shape you want.

**Comment by bobzwik** [comment]
Yeah, I'm not really understanding what the returned `r` is. I see that is is used in the `frag` function at line  66.https://github.com/vispy/vispy/blob/5e2985a8a0ec34e2353a1885cc4eb39ab97c0aec/vispy/visuals/markers.py#L47-L77

I don't see where `gl_PointCoord` comes from?

And I found the same transform (as my previous comment) used in the `vert` function, at line 40:
https://github.com/vispy/vispy/blob/5e2985a8a0ec34e2353a1885cc4eb39ab97c0aec/vispy/visuals/markers.py#L18-L44 
Could this be the source of `gl_PointCoord`?

I have to find a way to get vertices position from point coordinates, and then draw faces and edges. This is getting complicated fast!

**Comment by djhoese** [comment]
`gl_PointCoord` is being provided by the vertex shader's `gl_Position`...essentially. It's an OpenGL thing. I would suggest reading about it here (https://www.khronos.org/opengl/wiki/Built-in_Variable_(GLSL)) and seeing where that takes you.

For the `$transform`, yes exactly. The vertex shader is dealing with the "where" of things being drawn. The fragement shader is dealing with "what" things are drawn (think the color of the pixel on the screen). The `marker` function you would create would be called by the fragment shader here: https://github.com/vispy/vispy/blob/5e2985a8a0ec34e2353a1885cc4eb39ab97c0aec/vispy/visuals/markers.py#L66

If you copied and pasted this and modified it to deal with 3D you *might* get it to work, but with OpenGL there's always a chance of surprises.

**Comment by djhoese** [comment]
So thinking about it more, the marker functions are all essentially specifying how big the radius of the marker is. I'm not exactly sure how non-circular shapes are drawn yet.

**Comment by bobzwik** [comment]
> So thinking about it more, the marker functions are all essentially specifying how big the radius of the marker is. I'm not exactly sure how non-circular shapes are drawn yet.

Ahh I thought that is what it was doing, and I didn't understand how the shapes were actually made.

**Comment by djhoese** [comment]
If I'm being honest, I haven't written shader code like this before where it isn't a 1:1 coordinate:pixel color relationship. I think the idea here is that `gl_PointCoord` is essentially the coordinate (0 to 1 range for each dimension) of how far away from the vertex coordinate we are. So a lot of the math in the marker functions is to figure out how far away we are from the vertex coordinate and then the rest of the fragment shader sets the color based on that distance.

**Comment by bobzwik** [comment]
Ahhh I see. And I imagine that the frag function is looped for multiple `gl_PointCoord` between (0,0) and (1,1). I don't think I'll be able to use this method, but I'll try seeing what I can do by using code from `mesh.py`.

Thanks!

**Comment by djhoese** [comment]
Yes and keep in mind this isn't a vispy thing, this is an OpenGL thing. I would recommend googling how OpenGL works as far as vertex shaders and fragment shaders and the concepts involved for what they work on. "Looped" is not necessarily the right word since we're talking about GPUs and a lot of this processing is happening in parallel for every pixel/fragment. If you can get this to work it should perform really well (and selfishly, this would be a great addition to vispy).

What do you plan on pulling out of mesh.py?

**Comment by bobzwik** [comment]
I'd love to be able to contribute! But essentially, the vert and frag code in mesh.py is relatively simple, so  if I'm able to create vertices and faces for all the cubes (maybe using ```create_box()``` from `generation.py`) and just use those 2 functions, it should be possible to draw everything with just one gl program (?).

I'm getting my inspiration from work I'm doing in my research lab, and we often use rviz. We're able to flawlessly display pointclouds using markers, but also cubes and spheres. A colleague of mine suggested I look into how it's implemented in rviz (which uses OGRE to interface with openGL). So now I'm really getting lost into C++ territory.

**Comment by djhoese** [comment]
I'm not sure I can guarantee that it will work or work exactly the way you expect, but it is worth a shot. However, make sure you understand how the vertex shader and fragment shader are each running. It isn't like a normal C program where you might use for loops to go through each vertex or pixel. As long as you know that going in, then :+1: 

Very interesting. I've never used rviz or OGRE but I'm sure they are probably abstracting things away well enough that you'll run in to at least one unexpected headache when applying the same ideas to VisPy.

**Comment by bobzwik** [comment]
Before I try going the Markers way, I would like trying to go a different way : creating a new class BoxMarkersVisual. It would ressemble BoxVisual, but I would do just many `create_box()` and add them to a list of vertices. And then create a mesh.

However, how do I go about wrtiting `BoxMarkers = create_visual_node(visuals.BoxMarkersVisual)`? Shall I just make a copy of `vispy.scene.visuals`, edit it and import that one?

Thanks

**Comment by djhoese** [comment]
Assuming you are not adding this to a fork of the vispy code and this will live in your own codebase:

1. Create your BoxMarkersVisual class in your own python module/package.
2. Just below that, in the same module if you want, do `BoxMarkers = create_visual_node(BoxMarkersVisual)`.
3. When you want to use the visual with a SceneCanvas, import and use the `BoxMarkers` class.

**Comment by bobzwik** [comment]
Alright, well... success I guess! And it was easier than I expected. I basically made BoxMarkers and BoxMarkersVisual classes based off the BoxVisual class. I create a unit box, then scale it to the desired size and translate is for every point in  the pointcloud, So instead of ending up with a mesh of 24 vertices, I end up with a mesh of 24*n vertices, where `n` is the number of points. The result : instantaneous generation of the mesh (well, a slight (<1s) delay when generating over 100 000 boxes), and perfect frame performance (> 60 fps). [Here is a video of the result](https://www.youtube.com/watch?v=PfVFVJJ1nE4).

And here is my code : 
```# -*- coding: utf-8 -*-
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
import numpy as np
from vispy.geometry import create_box
from vispy.visuals.mesh import MeshVisual
from vispy.visuals.visual import CompoundVisual
from vispy.scene.visuals import create_visual_node

class BoxMarkersVisual(CompoundVisual):
    """Visual that displays a box.

    Parameters
    ----------
    point_coords : array_like
        Marker coordinates
    width : float
        Box width.
    height : float
        Box height.
    depth : float
        Box depth.
    width_segments : int
        Box segments count along the width.
    height_segments : float
        Box segments count along the height.
    depth_segments : float
        Box segments count along the depth.
    planes: array_like
        Any combination of ``{'-x', '+x', '-y', '+y', '-z', '+z'}``
        Included planes in the box construction.
    vertex_colors : ndarray
        Same as for `MeshVisual` class. See `create_plane` for vertex ordering.
    face_colors : ndarray
        Same as for `MeshVisual` class. See `create_plane` for vertex ordering.
    color : Color
        The `Color` to use when drawing the cube faces.
    edge_color : tuple or Color
        The `Color` to use when drawing the cube edges. If `None`, then no
        cube edges are drawn.
    """

    def __init__(self, point_coords=np.array([0,0,0]), width=1, height=1, depth=1, width_segments=1,
                 height_segments=1, depth_segments=1, planes=None,
                 vertex_colors=None, face_colors=None,
                 color=(0.5, 0.5, 1, 1), edge_color=None, **kwargs):
        
        self.point_coords = point_coords
        
        # Create a unit box
        width_box  = 1
        height_box = 1
        depth_box  = 1

        scale = np.array([width/width_box, height/height_box, depth/depth_box])

        vertices_box, filled_indices_box, outline_indices_box = create_box(
                width_box, height_box, depth_box, width_segments, height_segments,
                depth_segments, planes)

        # Store number of vertices, filled_indices and outline_indices per box
        self.nb_v  = vertices_box.shape[0]
        self.nb_fi = filled_indices_box.shape[0]
        self.nb_oi = outline_indices_box.shape[0]

        # Create empty arrays for vertices, filled_indices and outline_indices
        vertices = np.zeros(24*point_coords.shape[0],
                        [('position', np.float32, 3),
                         ('texcoord', np.float32, 2),
                         ('normal', np.float32, 3),
                         ('color', np.float32, 4)])
        filled_indices = np.zeros([12*point_coords.shape[0], 3], np.uint32)
        outline_indices = np.zeros([24*point_coords.shape[0], 2], np.uint32)

        # Iterate for every marker
        for i in range(point_coords.shape[0]):
            idx_v_start  = self.nb_v*i
            idx_v_end    = self.nb_v*(i+1)
            idx_fi_start = self.nb_fi*i
            idx_fi_end   = self.nb_fi*(i+1)
            idx_oi_start = self.nb_oi*i
            idx_oi_end   = self.nb_oi*(i+1)

            # Scale and translate unit box
            vertices[idx_v_start:idx_v_end]['position'] = vertices_box['position']*scale + point_coords[i]
            filled_indices[idx_fi_start:idx_fi_end] = filled_indices_box + idx_v_start
            outline_indices[idx_oi_start:idx_oi_end] = outline_indices_box + idx_v_start

        # Create MeshVisual for faces and borders
        self._mesh = MeshVisual(vertices['position'], filled_indices,
                                vertex_colors, face_colors, color, shading=None)
        if edge_color:
            self._border = MeshVisual(vertices['position'], outline_indices,
                                      color=edge_color, mode='lines')
        else:
            self._border = MeshVisual()

        CompoundVisual.__init__(self, [self._mesh, self._border], **kwargs)
        self.mesh.set_gl_state(polygon_offset_fill=True,
                               polygon_offset=(1, 1), depth_test=True)

    @property
    def mesh(self):
        """The vispy.visuals.MeshVisual that used to fill in.
        """
        return self._mesh

    @mesh.setter
    def mesh(self, mesh):
        self._mesh = mesh

    @property
    def border(self):
        """The vispy.visuals.MeshVisual that used to draw the border.
        """
        return self._border

    @border.setter
    def border(self, border):
        self._border = border


BoxMarkers = create_visual_node(BoxMarkersVisual)
```
I have a feeling there's a lot of stuff missing for a PR, what do you think?

For my own purposes, I am going to add methods to change the size and color of certain boxes (knowing the indices of the point, I can find the indices of the associated vertices). And I think `MeshData` already has methods of changing vertices positions and face colors (although they require position and color arrays for the whoooole mesh, and not just the modified ones). If you think it's pertinent, I can share the methods I create.

Transparency is a bit fiddly (same issues as with the simple Box class). I can post a video if you don't know about this issue yet. It gets "solved" if I change the `depth_test` of `self.mesh.set_gl_state` to `False`, but boxes end up dissapearing because openGL is rendering the edges of all the boxes far in the distance over the transparent faces.

**Comment by djhoese** [comment]
For any issues you are running in to I would ask that you look through the existing issues here on GitHub and see if you can find someone mentioning the same problem. As far as updating MeshData you may want to look at existing pull requests where @asnt has a lot of pending changes for Mesh handling, not all of them completely ready yet.

Drawing can get difficult when you have to play with different depth tests and other GL state parameters. Some visuals depend on the specific configuration they set and sometimes that doesn't work well with other visuals. Basic examples in an issue are welcome.
